java/src/main/c++/jni_spark_vw.cc

```diff
@@ -505,10 +596,145 @@ JNIEXPORT jobject JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitExample_predic
     // as this is not a ring-based example it is not free'd
     VW::LEARNER::as_singleline(all->l)->finish_example(*all, *ex);
 
-    return Java_org_vowpalwabbit_spark_VowpalWabbitExample_getPrediction(env, exampleObj);
+    return getJavaPrediction(env, all, ex);
+  }
+  catch (...)
+  {
+    rethrow_cpp_exception_as_java_exception(env);
+    // return null
+    return 0;
+  }
+}
+
+JNIEXPORT jstring JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitExample_toString(JNIEnv* env, jobject exampleObj)
+{
+  INIT_VARS
+
+  try
+  {
+    std::ostringstream ostr;
+
+    ostr << "VowpalWabbitExample(label=";
+    auto lp = all->p->lp;
+
+    if (!memcmp(&lp, &simple_label, sizeof(lp)))
+    {
+      label_data* ld = (label_data*)&ex->l;
+      ostr << "simple " << ld->label << ":" << ld->weight << ":" << ld->initial;
+    }
+    else if (!memcmp(&lp, &CB::cb_label, sizeof(lp)))
+    {
+      CB::label* ld = (CB::label*)&ex->l;
+      ostr << "CB " << ld->costs.size();
+
+      if (ld->costs.size() > 0)
+      {
+        ostr << " ";
+
+        CB::cb_class& f = ld->costs[0];
+
+        // ignore validation of action which is uniform_hash("shared")
+        if (f.partial_prediction == 0 && f.cost == FLT_MAX && f.probability == -1.f)
+          ostr << "shared";
+        else
+          ostr << f.action << ":" << f.cost << ":" << f.probability;
+      }
+    }
+    else
+    {
+      ostr << "unsupported label";
+    }
+
+    ostr << ";";
+    for (auto& ns : ex->indices)
+    {
+      if (ns == 0)
+        ostr << "NULL:0,";
+      else
+      {
+        if ((ns >= 'a' && ns <= 'z') || (ns >= 'A' && ns <= 'Z'))
+          ostr << "'" << (char)ns << "':";
+
+        ostr << (int)ns << ",";
+      }
+
+      for (auto& f : ex->feature_space[ns])
+      {
+        auto idx = f.index();
+        ostr << (idx & all->weights.mask()) << "/" << idx << ":" << f.value() << ", ";

```

Comment:
I'm guessing we're resilient to ending it with an extra comma?

Reply:
its just a debug string so its okay